File: tools/diagnostics/hs274-helper-environment.lua
-- tools/diagnostics/hs274-helper-environment.lua
-- Exercise the production environment boundary through native Hammerspoon tasks.

local root = assert(debug.getinfo(1, "S").source:match("^@(.+)/[^/]+$"))
local config = assert(hs.json.read(root .. "/helper-environment-config.json"))
local driver = config.repo .. "/static/ergopti_plus/macos"
package.path = driver .. "/?.lua;" .. driver .. "/?/init.lua;" .. package.path
local result = { kind = "native-helper-task-environment", bootstrap_verified = false }
local owner = {}
_G.hs274_helper_environment = owner

local function publish()
	assert(hs.json.write(result, config.result, true, true), "Cannot publish native environment result")
end

local function guarded(callback)
	return function(...)
		local ok, detail = xpcall(callback, debug.traceback, ...)
		if not ok then result.error = tostring(detail); publish() end
	end
end

local function run()
	assert(os.getenv("ERGOPTI_LAUNCHER_DEVICE") == nil, "Parent unexpectedly owns a launcher identity")
	assert(os.getenv("ERGOPTI_LAUNCHER_INODE") == nil, "Parent unexpectedly owns a launcher identity")
	result.parent_identity_absent = true
	-- The fixture isolates logging only; task construction and environment APIs are real.
	local logger = {}
	for _, name in ipairs({ "debug", "info", "warn", "error", "start", "success" }) do logger[name] = print end
	package.loaded["infra.logger"] = logger
	local runner = require("adapters.shell_runner")
	local function rejected_identity()
		local environment = {}
		for key, value in pairs(config.environment) do environment[key] = value end
		environment.ERGOPTI_LAUNCHER_INODE = config.wrong_inode
		owner.rejected = runner.spawn(config.helper, { "--remap-guardian-status" }, guarded(function(code, stdout)
			assert(code == 64 and stdout == "", "Wrong helper identity was admitted")
			result.wrong_identity_exit = code
			local refused = runner.spawn("/usr/bin/true", {}, nil, nil, { ERGOPTI_LOG_TOKEN = "fixture-only" })
			assert(refused.start() == false, "Launcher logger authority reached a child")
			result.forbidden_started = false
			publish()
		end), nil, environment)
		assert(owner.rejected.start(), "Identity rejection probe did not start")
	end
	owner.accepted = runner.spawn(config.helper, { "--remap-guardian-status" }, guarded(function(code, stdout, stderr)
		assert(code == 0, "Explicit native identity failed: " .. tostring(stderr))
		local status = stdout:match("^%s*(.-)%s*$")
		assert(status == "ready" or status == "requires_approval" or status == "unavailable", "Invalid native status")
		result.good_exit, result.guardian_status = code, status
		rejected_identity()
	end), nil, config.environment)
	assert(owner.accepted.start(), "Explicit native helper did not start")
end

guarded(run)()

File: tools/diagnostics/hs274_helper_environment.py
# tools/diagnostics/hs274_helper_environment.py
"""Reuse a verified helper to exercise real Hammerspoon child environments."""

import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import time

from hs274_hammerspoon import native_lifecycle
from hs274_hammerspoon_registration import register_application
from macos_native_helper import digest, source_identity

FIXTURE = Path(__file__).with_name("fixtures") / "hs274-native-helper-package.json"


def selection():
    """Load the retained native package rather than rebuilding unchanged Swift."""
    return json.loads(FIXTURE.read_text(encoding="utf-8"))


def authenticate(directory):
    """Require exact native receipt, archive bytes and current shared source inputs."""
    expected = selection()["identity"]
    directory = Path(directory)
    archives = list(directory.rglob("ErgoptiPlus.app.zip"))
    receipts = list(directory.rglob("native-helper.json"))
    if len(archives) != 1 or len(receipts) != 1:
        raise ValueError("Expected exactly one retained helper archive and receipt")
    archive, receipt = archives[0], receipts[0]
    if archive.is_symlink() or receipt.is_symlink():
        raise ValueError("Redirected native helper artifact")
    if json.loads(receipt.read_text(encoding="utf-8")) != expected:
        raise ValueError("Helper artifact differs from the accepted native receipt")
    if digest(archive) != expected["archive_sha256"] or archive.stat().st_size != expected["archive_bytes"]:
        raise ValueError("Helper archive integrity failed")
    if source_identity() != expected["sources"]:
        raise ValueError("Current native sources differ from the retained helper")
    return archive


def observe(directory, hammerspoon, output):
    """Run the production task adapter in an isolated real Hammerspoon process."""
    if sys.platform != "darwin" or os.environ.get("RUNNER_ENVIRONMENT") != "github-hosted":
        raise RuntimeError("Native environment acceptance requires a hosted macOS runner")
    archive = authenticate(directory)
    output = Path(output).absolute()
    output.mkdir()
    native_root = output / "runtime"
    native_root.mkdir()
    subprocess.run(["/usr/bin/ditto", "-x", "-k", str(archive), str(native_root)], check=True)
    helper_app = native_root / "ErgoptiPlus.app"
    helper = helper_app / "Contents/MacOS/ErgoptiPlus"
    subprocess.run(["/usr/bin/codesign", "--verify", "--strict", "--deep", str(helper_app)], check=True)
    if digest(helper) != selection()["identity"]["executable_sha256"]:
        raise ValueError("Extracted native helper executable changed")
    copied = native_root / "Hammerspoon.app"
    subprocess.run(["/usr/bin/ditto", str(hammerspoon), str(copied)], check=True)
    report = {}
    register_application(report, copied)
    here = Path(__file__).resolve().parent
    init = native_root / "init.lua"
    shutil.copyfile(here / "hs274-helper-environment.lua", init)
    executable = copied / "Contents/MacOS/Hammerspoon"
    result = output / "result.json"
    identity = helper.stat()
    config = {"repo": str(here.parents[1]), "helper": str(helper), "result": str(result),
              "wrong_inode": str(identity.st_ino + 1), "environment": {
                  "ERGOPTI_LAUNCHER_EXECUTABLE": str(helper), "ERGOPTI_LAUNCHER_DEVICE": str(identity.st_dev),
                  "ERGOPTI_LAUNCHER_INODE": str(identity.st_ino)}}
    (native_root / "helper-environment-config.json").write_text(json.dumps(config), encoding="utf-8", newline="\n")
    environment = os.environ.copy()
    for name in config["environment"]:
        environment.pop(name, None)
    lifecycle = native_lifecycle()
    native = lifecycle.NativeProcesses()
    with (output / "launch.log").open("xb") as log:
        launcher = subprocess.Popen(["/usr/bin/open", "-n", "-g", "-W", str(copied), "--args", "-MJConfigFile", str(init)],
                                    env=environment, stdout=log, stderr=subprocess.STDOUT)
        try:
            deadline = time.monotonic() + 20
            while not result.exists():
                if launcher.poll() is not None or time.monotonic() >= deadline:
                    raise RuntimeError("Native Hammerspoon environment probe did not settle")
                time.sleep(0.05)
            receipt = json.loads(result.read_text(encoding="utf-8"))
            if (receipt.get("error") or receipt.get("good_exit") != 0 or receipt.get("wrong_identity_exit") != 64
                    or receipt.get("parent_identity_absent") is not True or receipt.get("forbidden_started") is not False
                    or len(native.matching(executable)) != 1):
                raise RuntimeError("Native task environment acceptance failed: " + repr(receipt))
            report["result"] = receipt
        finally:
            lifecycle.cleanup(native, executable, launcher)
            report["application_cleanup"] = "confirmed"
            (output / "receipt.json").write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8", newline="\n")


if __name__ == "__main__":
    if sys.argv[1:] == ["selection"]:
        selected = selection()
        print("run_id=" + str(selected["producer_run"]))
        print("artifact_name=" + selected["artifact_name"])
    elif len(sys.argv) == 4:
        observe(*sys.argv[1:])
    else:
        raise SystemExit("Expected selection, or artifact directory, Hammerspoon app, and new output directory")
